PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/DeviceLoginWindow.java

```diff
@@ -0,0 +1,139 @@
+        setTitle(TITLE);
+        editorPanel.setText(createHtmlFormatMessage(deviceCode.getMessage()));
+        editorPanel.addHyperlinkListener(e -> {
+            if (e.getEventType() == HyperlinkEvent.EventType.ACTIVATED) {
+                try {
+                    Desktop.getDesktop().browse(e.getURL().toURI());
+                } catch (IOException | URISyntaxException e1) {
+                    // swallow exceptions
+                }
+            }
+        });
+        authExecutor = ApplicationManager.getApplication()
+            .executeOnPooledThread(() -> pullAuthenticationResult(ctx, deviceCode, callBack));
+        init();
+    }
+
+    private void pullAuthenticationResult(final AuthenticationContext ctx, final DeviceCode deviceCode,
+                                          final AuthenticationCallback<AuthenticationResult> callback) {
+        final long interval = deviceCode.getInterval();
+        long remaining = deviceCode.getExpiresIn();
+        while (remaining > 0 && authenticationResult == null) {
+            try {
+                remaining -= interval;
+                Thread.sleep(interval * 1000);
+                authenticationResult = ctx.acquireTokenByDeviceCode(deviceCode, callback).get();
+            } catch (ExecutionException | InterruptedException e) {
+                if (e.getCause() instanceof AuthenticationException &&
+                    ((AuthenticationException) e.getCause()).getErrorCode() == AdalErrorCode.AUTHORIZATION_PENDING) {
+                    // swallow the pending exception
+                } else {
+                    e.printStackTrace();
+                    break;
+                }
+            }
+        }
+        closeDialog();
+    }
+
+    private String createHtmlFormatMessage(final String message) {
+        final String pattern = "(https?)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]";
+        final Matcher matcher = Pattern.compile(pattern).matcher(message);
+        if (matcher.find()) {
+            final String deviceLoginUrl = matcher.group(0);
+            return "<p>"
+                + message.replace(deviceLoginUrl, String.format("<a href=\"%s\">%s</a>", deviceLoginUrl,
+                deviceLoginUrl))
+                + "</p><p>Waiting for signing in with the code ...</p>";
+        }
+        return message;
+    }
+
+    @Override
+    public void doCancelAction() {
+        authExecutor.cancel(true);
+        super.doCancelAction();
+    }
+
+    private void closeDialog() {
+        ApplicationManager.getApplication().invokeLater(() -> {
+            final Window w = getWindow();
+            w.dispatchEvent(new WindowEvent(w, WindowEvent.WINDOW_CLOSING));
+        }, ModalityState.stateForComponent(jPanel));
+    }
+
+    @Override
+    protected Action[] createActions() {
+        return new Action[]{this.getCancelAction()};
+    }
+
+    @Nullable
+    @Override
+    protected JComponent createCenterPanel() {
+        return jPanel;
```

Comment:
If you program with Kotlin, I'd like to suggest you have a try of our new introduced form builder DSL `FormBuilder`. 😃

Reply:
Haven't got the change to program with Kotlin. Sounds good.